### ISSUE TITLE
how to test model in real world?

### ISSUE THREAD

**Issue by nikkisingh111333** [oracle/tribuo#87]
hey,i have currently tested random forest and everything is good for now.But how can i start predicting i.e take inputs from user and then give them the result modified .I was digging api and found Example class but dont know how make it a feature of inputs to use in model.predict function.for e.g if i want to predict height using feature age and weight from user how would i do that .i was considering to use datasource class and finding out something which can turn list of inputs into Example dataset then passing it to model.predict() ..is there any Tribuo way i should use?

**Comment by Craigacp** [comment]
If you used a `ColumnarDataSource` then you can get the row processor out and use that to convert a `Map<String,String>` into an `Example`. This is covered in the columnar data tutorial 
https ://github.com/oracle/tribuo/blob/main/tutorials/columnar-tribuo-v4.ipynb.

Otherwise it depends on where your inference data is. If it's streaming in then you'll need to build the `Example`s and `Feature`s from the data. Use `ArrayExample` in preference to `ListExample`, and prefer to build up a `List<Feature>` and add it to the example in one go. Use `OutputFactory.getUnknownOutput()` to generate an output if you don't know the ground truth label. If your data is on disk then load it using the same mechanism that you loaded the training data.

**Comment by nikkisingh111333** [comment]
hey i m using coulumnar api and getting this Error everytime.
`WARNING: RowProcessor was already configured, yet expandRegexMapping was called with [PetalWidthCm, PetalLengthCm, SepalWidthCm, SepalLengthCm]
Nov 16, 2020 1:28:54 AM org.tribuo.data.columnar.RowProcessor generateExample
WARNING: Row 0 empty of features, omitting
Exception in thread "main" java.util.NoSuchElementException: No value present
	at java.util.Optional.get(Optional.java:135)
	at MachineLearningWithTRIBUO.TribuoML.DecisionTree(TribuoML.java:114)
	at MachineLearningWithTRIBUO.TribuoML.main(TribuoML.java:229)
`
here my code copied from api
`      ArrayList<FieldExtractor<?>> metadataExtractors = new ArrayList<FieldExtractor<?>>();
       metadataExtractors.add(new IdentityExtractor("PetalWidthCm"));
       metadataExtractors.add(new IdentityExtractor("SepalWidthCm"));
       metadataExtractors.add(new IdentityExtractor("PetalLengthCm"));
       metadataExtractors.add(new IdentityExtractor("SepalLengthCm"));
       HashMap regexMappingProcessors = new HashMap<String,FieldProcessor>();
       RowProcessor rowProcessor = new RowProcessor<Label>(metadataExtractors,null,new FieldResponseProcessor("Species","UNK",new LabelFactory()),new HashMap<String,FieldProcessor>(),Collections.emptySet());
       Map<String,String> newRow= ImmutableMap.of("PetalWidthCm", "5", "PetalLengthCm", "5.6", "SepalWidthCm", "8","SepalLengthCm","7");
        System.out.println(newRow);
        List<String> headers=Collections.unmodifiableList(new ArrayList<>(newRow.keySet()));
        Row row = new ColumnarIterator.Row(0,headers,newRow);
        System.out.println(row);
        rowProcessor.expandRegexMapping(headers);
        Example<Label> example = (Example<Label>) rowProcessor.generateExample(row,false).get();
        example.toString();`
now what ? kinda stuck

**Comment by Craigacp** [comment]
The error message from the `RowProcessor` says it failed to find any features, because no `FieldProcessor`s were configured. You've configured it to extract all the values as metadata (using `FieldExtractor`) not field processors. You want to use `org.tribuo.data.columnar.processors.field.IdentityProcessor` not `org.tribuo.data.columnar.extractors.IdentityExtractor`, then supply them to the 4th argument of the `RowProcessor` constructor which is currently getting an empty `HashMap`. For the standard irises dataset there aren't any metadata fields so you can use the simpler `RowProcessor` constructor - `public RowProcessor(ResponseProcessor<T> responseProcessor, Map<String,FieldProcessor> fieldProcessorMap)`.

**Comment by nikkisingh111333** [comment]
After using above instruction i m getting this error now 
`Exception in thread "main" java.lang.IllegalArgumentException: No features found in Example ArrayExample(numFeatures=4,output=SPECIES,weight=1.0,features=[(PetalLengthCm@value, 5.6)(PetalWidthCm@value, 5.8), (SepalLengthCm@value, 7.0), (SepalWidthCm@value, 8.0), ])`
Heres my code:
`       Map<String,FieldProcessor> fieldProcessors = new HashMap<String,FieldProcessor>();
       fieldProcessors.put("PetalWidthCm",new DoubleFieldProcessor("PetalWidthCm"));
       fieldProcessors.put("PetalLengthCm",new DoubleFieldProcessor("PetalLengthCm"));
       fieldProcessors.put("SepalLengthCm",new DoubleFieldProcessor("SepalLengthCm"));
       fieldProcessors.put("SepalWidthCm",new DoubleFieldProcessor("SepalWidthCm"));
       RowProcessor rowProcessor = new RowProcessor<Label>(new FieldResponseProcessor("Species","Species",new LabelFactory()),fieldProcessors);
       Map<String,String> newRow= ImmutableMap.of("Id","2","PetalWidthCm", "5.8", "PetalLengthCm", "5.6", "SepalWidthCm",    "8","SepalLengthCm","7");
        System.out.println(newRow);
        List<String> headers=Collections.unmodifiableList(new ArrayList<>(newRow.keySet()));
        System.out.println(headers);
        Row row = new ColumnarIterator.Row(0,headers,newRow);
        System.out.println(row);
        rowProcessor.expandRegexMapping(headers);
        Map output=model.predict((Example) rowProcessor.generateExample(row,false).get()).getOutputScores();
       System.out.println( "Rumba:"+rowProcessor.generateExample(row,false).get());
       System.out.println("Output:"+output);`

now what? its also transforming feature name with @value like this:
`ArrayExample(numFeatures=4,output=SPECIES,weight=1.0,features=[(PetalLengthCm@value, 5.6)(PetalWidthCm@value, 5.8), (SepalLengthCm@value, 7.0), (SepalWidthCm@value, 8.0), ])
WHAT DOSS IT MEAN?
`

**Comment by Craigacp** [comment]
You should try to use the same input processing pipeline at training and test time. The error is because `CSVLoader` uses slightly different feature names to the `IdentityProcessor` for numerical features. If you're using the `RowProcessor` at test time then use `CSVDataSource` at training time to load in irises, and pull out the `RowProcessor` from the training data source to use at test time. This is described in the [columnar-data-tutorial](https://github.com/oracle/tribuo/blob/main/tutorials/columnar-tribuo-v4.ipynb). Though given it's the iris dataset you could build the features and examples by hand as I suggested earlier.

**Comment by nikkisingh111333** [comment]
I m using csvdatasource but not working ..how can i pull RowProcessor from
datasource its a kind of mess.

**Comment by Craigacp** [comment]
You have to pass the `RowProcessor` into `CSVDataSource` on construction. Hold on to a reference to it when you do.

**Comment by nikkisingh111333** [comment]
oKK IT worked actually i was using it reversely now its giving me data
result...few more questions please?
1.whats the main diffrence between csvloader and csvdatasource i mean when
to use what if i again stuck on same kind of problem?
2.what about data with more than 50 features do i need to create
map<string,string> for every features dataset like minist or any other
flatten imagedata?
3.how to use tribou with opencv as i m also in opencv stuffs like
facedetections handdetection etc .how can i transform image dataset into
flattened Array of features in python it use ndarray or vectorize is there
anything to offer by tribou or in future?

**Comment by Craigacp** [comment]
`CSVLoader` is for extremely simple numerical csv files with a single response column, and all other columns are treated as features. It's used for basic demos and other things which fit in the parameters above. If you need to do anything else use `CSVDataSource`. The tutorials in 4.0.2 have been updated to better describe the differences between the two options.

If you have lots of features you want to process the same way then you can use the regex mapping, find a regex that matches all the columns and supply that. Again this is covered in the columnar tutorial. For MNIST you can use the `IDXDataSource` which loads it in the original format.

Tribuo currently has little support for image recognition tasks as we've been focusing on NLP and tabular problems. For images it's best if you write your own `DataSource` and have it flatten the images out into Tribuo's feature representation. Then you can pass them to the `TensorFlowTrainer` or the `ONNXExternalModel` and use the image processors there to convert it back into a tensor format. This is suboptimal, and we know that, but we've not put much effort into making image tasks work well yet. Part of the reason is that the sparse assumption that Tribuo makes about it's data isn't true for image recognition tasks, and so we'll need to expand the type system to more efficiently work with dense data. At the moment for image recognition tasks on the JVM you might want to look at using [Tensorflow Java](https://github.com/tensorflow/java) directly, or one of the other Java deep learning interfaces.

**Comment by nikkisingh111333** [comment]
thanks for this .so basically tribuo currently is for simple taabular data .and we can use little minst kind of stuffs for now.ok i understand ..thanks for your support.
